### PR TITLE
Upgrade nits

### PR DESCRIPTION
**Pull request by aaronbuchwald**

### CODE REVIEW

File: core/genesis.go [hunk 1]
```diff
@@ -350,13 +356,17 @@ func (g *Genesis) Commit(db ethdb.Database) (*types.Block, error) {
...
 		return nil, err
 	}
-	rawdb.WriteBlock(db, block)
-	rawdb.WriteReceipts(db, block.Hash(), block.NumberU64(), nil)
-	rawdb.WriteCanonicalHash(db, block.Hash(), block.NumberU64())
-	rawdb.WriteHeadBlockHash(db, block.Hash())
-	rawdb.WriteHeadHeaderHash(db, block.Hash())
-	rawdb.WriteChainConfig(db, block.Hash(), config)
-	rawdb.WriteUpgradeConfig(db, block.Hash(), &config.UpgradeConfig)
+	batch := db.NewBatch()
```
Comment (ceyonur) [inline core/genesis.go:359]: ooh that probably also goes to coreth + even geth?

Comment (aaronbuchwald) [reply core/genesis.go:359]: Ya I don't think we've ever a report of this happening, but there was a warn log in there to explicitly handle one case where the genesis block is written, but the chain config hasn't been written yet, which can be avoided by writing the block and chain config together on one batch.

**Review by ceyonur — APPROVED** [review]
noice

File: core/genesis.go [hunk 2]
```diff
@@ -350,13 +356,17 @@ func (g *Genesis) Commit(db ethdb.Database) (*types.Block, error) {
...
+	rawdb.WriteBlock(batch, block)
+	rawdb.WriteReceipts(batch, block.Hash(), block.NumberU64(), nil)
+	rawdb.WriteCanonicalHash(batch, block.Hash(), block.NumberU64())
+	rawdb.WriteHeadBlockHash(batch, block.Hash())
+	rawdb.WriteHeadHeaderHash(batch, block.Hash())
+	rawdb.WriteChainConfig(batch, block.Hash(), config)
+	rawdb.WriteUpgradeConfig(batch, block.Hash(), &config.UpgradeConfig)
+	if err := batch.Write(); err != nil {
+		return nil, fmt.Errorf("failed to write genesis block: %w", err)
+	}
```
Comment (darioush) [inline core/genesis.go:369]: nice :)

File: params/precompile_config.go [hunk 3]
```diff
@@ -62,14 +63,21 @@ func (c *ChainConfig) VerifyPrecompileUpgrades() error {
...
+			config, ok := upgrade.getByKey(key)
+			if !ok {
 				continue
 			}
 			if hasKey {
 				return fmt.Errorf("PrecompileUpgrades[%d] has more than one key set", i)
 			}
+			if config.Timestamp() == nil {
+				return fmt.Errorf("PrecompileUpgrades[%d] cannot have a nil timestamp", i)
+			}
```
Comment (darioush) [inline params/precompile_config.go:75]: good idea

File: params/precompile_config.go [hunk 4]
```diff
@@ -62,14 +63,21 @@ func (c *ChainConfig) VerifyPrecompileUpgrades() error {
...
 			hasKey = true
 		}
+		if !hasKey {
+			return fmt.Errorf("empty precompile upgrade at index %d", i)
+		}
```
Comment (darioush) [inline params/precompile_config.go:80]: also a good idea

File: utils/fork.go [hunk 5]
```diff
@@ -22,3 +22,15 @@ func IsForkTransition(fork *big.Int, parent *big.Int, current *big.Int) bool {
...
+// contain the same value.
+func BigNumEqual(x, y *big.Int) bool {
+	if x == nil {
+		return y == nil
+	}
+	if y == nil {
+		return x == nil
+	}
+	return x.Cmp(y) == 0
+}
```
Comment (darioush) [inline utils/fork.go:36]: nice

**Review by darioush — APPROVED** [review]

**Comment by darioush** [comment]
(some tests need updating)

**Pull request merged**